src/org/sosy_lab/common/collect/SkipList.java

```diff
@@ -163,37 +166,86 @@ private void readObject(ObjectInputStream pIn) throws IOException, ClassNotFound
   private static final double ONE_HALF_LOG = Math.log(0.5);
   private Random randomGenerator = new Random();
 
-  private final Comparator<? super T> comparator;
+  private final @Nullable Comparator<? super T> comparator;
   private Node<T> head = createHead();
   private Node<T> tail = head;
   private int size = 0;
 
-  @SuppressWarnings("unchecked")
-  public SkipList(@Nullable Comparator<? super T> pComparator) {
-    if (pComparator == null) {
-      comparator = (Comparator<? super T>) Comparator.naturalOrder();
-    } else {
-      comparator = pComparator;
-    }
+  private SkipList(Comparator<? super T> pComparator) {
+    Preconditions.checkNotNull(pComparator);
+    comparator = pComparator;
   }
 
-  public SkipList(Collection<? extends T> pCollection) {
+  private SkipList(Iterable<? extends T> pCollection) {
     this();
-    boolean changed = addAll(pCollection);
-    assert changed;
+    addAll0(pCollection);
   }
 
-  public SkipList(SortedSet<T> pSortedSet) {
-    this(pSortedSet.comparator());
-    boolean changed = addAll(pSortedSet);
-    assert changed;
+  private SkipList(Iterable<? extends T> pCollection, Comparator<? super T> pComparator) {
+    this(pComparator);
+    addAll0(pCollection);
   }
 
-  @SuppressWarnings("unchecked")
-  public SkipList() {
-    comparator = (Comparator<? super T>) Comparator.naturalOrder();
+  private SkipList() {
+    comparator = null;
+  }
+
+  public static <T extends Comparable<T>> SkipList<T> of(T p1) {
+    SkipList<T> list = new SkipList<>();
+    list.add(p1);
+    return list;
+  }
+
+  public static <T extends Comparable<T>> SkipList<T> of(T p1, T p2) {
+    SkipList<T> list = new SkipList<>();
+    list.addAll(Lists.newArrayList(p1, p2));
+    return list;
+  }
+
+  public static <T extends Comparable<T>> SkipList<T> of(T p1, T p2, T p3) {
+    SkipList<T> list = new SkipList<>();
+    list.addAll(Lists.newArrayList(p1, p2, p3));
+    return list;
   }
 
+  @SafeVarargs
+  public static <T extends Comparable<T>> SkipList<T> of(T p1, T p2, T p3, T... pOtherElements) {
+    SkipList<T> list = new SkipList<>();
+    list.addAll(Lists.newArrayList(p1, p2, p3));
+    for (T v : pOtherElements) {
+      list.add(v);
+    }
+    return list;
+  }
+
+  public static <T extends Comparable<T>> SkipList<T> create() {
+    return new SkipList<>();
+  }
+
+  public static <T> SkipList<T> create(Comparator<? super T> pComparator) {
+    return new SkipList<>(pComparator);
+  }
+
+  public static <T> SkipList<T> create(
+      Iterable<? extends T> pCollection, Comparator<? super T> pComparator) {
+    return new SkipList<>(pCollection, pComparator);
+  }
+
+  public static <T extends Comparable<T>> SkipList<T> createWithNaturalOrder(
+      Iterable<? extends T> pCollection) {
+    return new SkipList<>(pCollection);
+  }
+
+  public static <T> SkipList<T> copyOf(SortedSet<T> pSortedSet) {
```

Comment:
How about `copyOfWithSameOrder`?

Or do you think the behavior of this method is what most people would expect from a `copyOf` method with no further explanation?

I just noted that Guava's `ImmutableSortedSet` has the reverse naming: `copyOf` uses natural order, `copyOfSorted` uses the same order. This is short and easy, and I think would be nice to adopt.

Reply:
Here I used `copyOf` because the elements and their order are both kept, so it is a real 'copy' (so also 'yes' to your second question).

But for factory methods, Guava actually seems to differentiate between `ImmutableXXX` and other collections (e.g., `TreeMultiset`). The second only provide `create(...)` methods.

So should we just align ourselves with the `ImmutableXXX` classes for all factory methods, to have a naming scheme consistent with Guava classes, since you mentioned that `of(…)` methods are useful?